diff --git a/Makefile.in b/Makefile.in
index 76f8c1e786..d3a285554a 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -110,6 +110,9 @@ am__v_at_1 =
 
 # Location to install Emacs.app on macOS
 mac_appdir=@mac_appdir@
+mac_appbindir=@mac_appbindir@
+mac_appresdir=@mac_appresdir@
+mac_self_contained=@mac_self_contained@
 
 # Location to install Emacs.app under GNUstep / macOS.
 # Later values may use these.
@@ -512,6 +515,10 @@ install-arch-dep:
 	subdir=${ns_appresdir}/site-lisp && ${write_subdir}
 	rm -rf ${ns_appresdir}/share
 endif
+ifeq (${mac_self_contained},yes)
+	subdir=${mac_appresdir}/site-lisp && ${write_subdir}
+	rm -rf ${mac_appresdir}/share
+endif
 
 ### Windows-specific install target for installing programs produced
 ### in nt/, and its Posix do-nothing shadow.
diff --git a/configure.ac b/configure.ac
index d2fb6fb328..c7a95287a5 100644
--- a/configure.ac
+++ b/configure.ac
@@ -436,6 +436,11 @@ AC_DEFUN
 		 [DIR=/Application]])],
 [ mac_appdir_x=${enableval}])
 
+AC_ARG_ENABLE(mac-self-contained,
+[AS_HELP_STRING([--enable-mac-self-contained],
+                [enable self contained build under mac])],
+                EN_MAC_SELF_CONTAINED=$enableval)
+
 AC_ARG_WITH([gnustep-conf],dnl
 [AS_HELP_STRING([--with-gnustep-conf=FILENAME],
    [name of GNUstep configuration file to use on systems where the command
@@ -1867,6 +1872,18 @@ AC_DEFUN
 if test "${with_mac}" != no; then
   AC_CHECK_HEADER(Carbon/Carbon.h, HAVE_MACGUI=yes)
   test "${HAVE_MACGUI}" = yes && window_system=mac
+
+  ## Specify the install directory
+  mac_appdir=`pwd`/mac
+  if test "${mac_appdir_x}" != ""; then
+    case ${mac_appdir_x} in
+      y | ye | yes)  mac_appdir=/Applications ;;
+      * ) mac_appdir=${mac_appdir_x} ;;
+    esac
+  fi
+
+  mac_appbindir=${mac_appdir}/Emacs.app/Contents/MacOS
+  mac_appresdir=${mac_appdir}/Emacs.app/Contents/Resources
 fi
 
 HAVE_NS=no
@@ -1980,14 +1997,34 @@ AC_DEFUN
 
 AC_SUBST(LIBS_GNUSTEP)
 
+mac_self_contained=no
 MAC_OBJ=
 MAC_OBJC_OBJ=
 if test "${HAVE_MACGUI}" = yes; then
+  if test "${EN_MAC_SELF_CONTAINED}" = yes; then
+     mac_self_contained=yes
+     prefix=${mac_appresdir}
+     exec_prefix=${mac_appbindir}
+     dnl This one isn't really used, only archlibdir is.
+     libexecdir="\${mac_appbindir}/libexec"
+     archlibdir="\${mac_appbindir}/libexec"
+     etcdocdir="\${mac_appresdir}/etc"
+     etcdir="\${mac_appresdir}/etc"
+     dnl FIXME maybe set datarootdir instead.
+     dnl That would also get applications, icons, man.
+     infodir="\${mac_appresdir}/info"
+     mandir="\${mac_appresdir}/man"
+     lispdir="\${mac_appresdir}/lisp"
+     test "$locallisppathset" = no && locallisppath=""
+     INSTALL_ARCH_INDEP_EXTRA=
+  fi
+
   MAC_OBJ="mac.o macterm.o macfns.o macmenu.o macselect.o"
   MAC_OBJC_OBJ="macappkit.o macfont.o"
 fi
 AC_SUBST(MAC_OBJ)
 AC_SUBST(MAC_OBJC_OBJ)
+AC_SUBST(mac_self_contained)
 
 INSTALL_ARCH_INDEP_EXTRA=install-etc
 ns_self_contained=no
@@ -3763,14 +3800,6 @@ AC_DEFUN
   AC_DEFINE(HAVE_SOUND, 1, [Define to 1 if you have sound support.])
   AC_CHECK_HEADERS(AvailabilityMacros.h)
   MAC_CFLAGS="-fconstant-cfstrings"
-  ## Specify the install directory
-  mac_appdir=
-  if test "${mac_appdir_x}" != ""; then
-    case ${mac_appdir_x} in
-      y | ye | yes)  mac_appdir=/Applications ;;
-      * ) mac_appdir=${mac_appdir_x} ;;
-    esac
-  fi
   # We also have mouse menus.
   HAVE_MENUS=yes
   # Tell src/Makefile.in to create files in the macOS application
@@ -5064,6 +5093,8 @@ AC_DEFUN
 ## Used in lwlib/Makefile.in.
 AC_SUBST(X_TOOLKIT_TYPE)
 AC_SUBST(mac_appdir)
+AC_SUBST(mac_appbindir)
+AC_SUBST(mac_appresdir)
 AC_SUBST(ns_appdir)
 AC_SUBST(ns_appbindir)
 AC_SUBST(ns_appresdir)
@@ -5523,6 +5554,13 @@ AC_DEFUN
 Use it at your own risk."
       echo
    fi
+   if test "$EN_MAC_SELF_CONTAINED" = "yes"; then
+      echo "The application will be fully self-contained."
+   else
+      AS_ECHO(["The lisp resources for the application will be installed under ${prefix}.
+You may need to run \"${MAKE-make} install\" with sudo.  The application will fail
+to run if these resources are not installed."])
+   fi
 fi
 
 if test "$HAVE_NS" = "yes"; then
diff --git a/lib-src/Makefile.in b/lib-src/Makefile.in
index 5b526f1bf5..54319f3c53 100644
--- a/lib-src/Makefile.in
+++ b/lib-src/Makefile.in
@@ -77,6 +77,7 @@ am__v_at_1 =
 # Location to install Emacs.app under GNUstep / macOS.
 # Later values may use this.
 ns_appbindir=@ns_appbindir@
+mac_appbindir=@mac_appbindir@
 
 # The default location for installation.  Everything is placed in
 # subdirectories of this directory.  The default values for many of
diff --git a/mac/Emacs.app/Contents/MacOS/Emacs.sh b/mac/Emacs.app/Contents/MacOS/Emacs.sh
index ef3ea5ccc1..b76d765fef 100755
--- a/mac/Emacs.app/Contents/MacOS/Emacs.sh
+++ b/mac/Emacs.app/Contents/MacOS/Emacs.sh
@@ -21,7 +21,7 @@
 ### Code:
 
 export EMACS_REINVOKED_FROM_SHELL=1
-set "${0%.sh}" "$@"
+[ -L $0 ] && set "$(readlink ${0%.sh})" "$@" || set "${0%.sh}" "$@"
 
 case $(sw_vers -productVersion) in
     10.[0-7]|10.[0-7].*)
